extension-bolt: taproot gossip (features 70/71/72/73/74/75) (#1059)
lightning/bolts: bolt-taproot-gossip.md

  * [TLV Based Messages](#tlv-based-messages)
  * [Taproot Channel Proof and Verification](#taproot-channel-proof-and-verification)
  * [Timestamp fields](#timestamp-fields)
  * [Bootstrapping Taproot Gossip](#bootstrapping-taproot-gossip)
  * [Specification](#specification)
    * [Feature Bits](#feature-bits)
    * [The `announcement_signatures_2` message](#the-announcement_signatures_2-message)
    * [The `channel_announcement_2` message](#the-channel_announcement_2-message)
    * [The `node_announcement_2` message](#the-node_announcement_2-message)
    * [The `channel_update_2` message](#the-channel_update_2-message)
  * [Appendix A: Algorithms](#appendix-a-algorithms)
      * [Partial Signature Calculation](#partial-signature-calculation)
      * [Partial Signature Verification](#partial-signature-verification)
  * [Appendix B: Test Vectors](#appendix-b-test-vectors)
  * [Acknowledgements](#acknowledgements)

## Aim 

This document aims to update the gossip protocol defined in [BOLT 7][bolt-7] to
allow for advertisement and verification of taproot channels. An entirely new
set of gossip messages are defined that use [BIP-340][bip-340] signatures and

> that use a purely TLV based schema.

Reviewer: You can't though, because that would not be backwards compatible :(  So you actually do have compulsory fields, which cannot be removed, until everyone upgrades.

It's actually simpler to define a new type in that case!

## Overview

The initial version of the Lightning Network gossip protocol as defined in
[BOLT 7][bolt-7] was designed around P2WSH funding transactions. For these
channels, the `channel_announcement` message is used to advertise the channel to
the rest of the network. Nodes in the network use the content of this message to
prove that the channel is sufficiently bound to the Lightning Network context
and that it is owned by the nodes advertising the channel. This proof and
verification protocol is, however, not compatible with SegWit V1 (P2TR) outputs
and so cannot be used to advertise the channels defined in the
[Simple Taproot Channel][simple-taproot-chans] proposal. This document thus aims
to define an updated gossip protocol that will allow nodes to both advertise and
verify taproot channels. This part of the update affects the
`announcement_signatures` and `channel_announcement` messages.

> The opportunity is also taken to rework the `node_announcement` and

Author: [From Joost](https://github.com/ellemouton/bolts/pull/2#discussion_r1135117400):

Could you also not take this opportunity and keep using ecdsa signatures for those? That seems to reduce the delta of this proposal significantly, and maybe without loss of functionality for the user?

Author: yes we could definitely do this. but: 
- getting to use schnorr sigs for them is quite beneficial for things like multi-key backed nodes
- see the section on "Bootstrapping Taproot Gossip", it is not necessary for nodes to use the new announcement message straight away. They can use the old one for as long as they like.

an announced taproot channel to their routing graph:

1. The funding output of the channel being advertised exists on-chain, is an
   unspent UTXO and has a sufficient number of confirmations. To allow this 
   check, the SCID of the channel will continue to be included in the channel
   announcement.
2. The funding transaction is sufficiently bound to the Lightning context. 
   Nodes will do this by checking that they are able to derive the output key 
   found on-chain by using the advertised `bitcoin_key_1` and `bitcoin_key_2` 
   along with a [BIP86 tweak][bip86-tweak]. This provides a slightly weaker
   binding to the LN context than legacy channels do but at least somewhat 
   limits how the output can be spent due to the script-path being disabled.
   The context binding with legacy channels is greater because an attacker 
   trying to create gossip spam would need to create 2-of-2 multisig P2WSH 
   outputs on-chain that would require them to then pay for the bytes of two
   signatures at the time of spending the output. This extra cost provided some
   extra deterrence for attackers. This deterrence is not present with P2TR 
   transactions.

> 3. The owners of `bitcoin_key_1` and `bitcoin_key_2` agree to be associated with  

Author: [From George](https://github.com/ellemouton/bolts/pull/2#discussion_r1135955383): 

Since the public announcements of channels pretty much reveal what the output is being used for, would it be fair to mention that this mainly benefits private channels?

Different phrasing: Taproot outputs all do look the same regardless of being used for a channel or not, but any participant in LN will eventually receive a (public) channel announcement that marks some taproot UTXO as a channel, so effectively it's only the private channels that can remain "incognito"

   a channel owned by `node_1` and `node_2`.
4. The owners of `node_ID_1` and `node_ID_2` agree to being associated with the 
   output paying to `bitcoin_key_1` and `bitcoin_key_2`. 

For legacy channels, these last two proofs are made possible by including four
separate signatures in the `channel_announcement`. However, [BIP-340][bip-340]
signatures and the MuSig2 protocol allow us to now aggregate these four
signatures into a single one. Verifiers will then be able to aggregate the four
keys (`bitcoin_key_1`, `bitcoin_key_2`, `node_ID_1` and `node_ID_2`) using
MuSig2 key aggregation and then they can do a single signature verification

> check instead of four individual checks.

Author: I think there is some benefit to proving a binding to the LN context at least a little bit. 

We could possibly just advertise the `taproot_internal_key` though (although I need to check the musig verification math for this). Kinda nice to, at least initially, to prove that the output cant be spent via the script path (via BIP86) as this at least gives spammers _some_ deterrence... although perhaps not enough - interested to hear what people think re preventing spam of non-channels in the "only advertise the taproot_output_key" case

> check instead of four individual checks.

Author: I think that perhaps the "only announce taproot_output_key" scenario only works in full Gossip 2.0 where you have to prove you have some bitcoin in order to announce channels. But in this proposal (gossip 1.5), I think that might not be the case.

> check instead of four individual checks.

Reviewer: Gossip v2 detached the proof-of-utxo from the channels in two ways:
1. You could use any UTXO proofs
2. The proofs were per-node, and not per-channel.

This is the former, without the latter.  It allows for us to change to a different Taproot scheme in future if we wanted, without updating gossip, but still requires a proof-per-channel.  Say, gossip 1.75?

## Timestamp fields

In this document, the `timestamp` fields in messages are block heights instead
of the UNIX timestamps used in the legacy gossip messages. 

### Rate Limiting

> A block height based timestamp results in more natural rate limiting for gossip

Reviewer 2: Pretty cool how we can indirectly gain global rate limiting by using block hashes. I think this has a few implications for the ways `channel_updates` work in the network today:
  * Any attempts to use `max_htlc` to leak out the available bandwidth for a channel (so increase/decrease with each forward) can be rate limited at a global level. 
  * Protocols that would directly require a node to trigger a specific channel update are now less attractive from an incentives PoV as: those updates would eat into the node's available set of global updates within a block interval. 
  * Nodes can _still_ do more rapid updates by sending their new `channel_update` in an onion error _directly_ to the sender. 

Also some implementations have extensions to the gossip query protocol to support timestamp channel update syncing, which can now be updated to use block height based syncing instead.

messages: nodes are allowed to send at most one announcement and update per
block. To allow for bursts, nodes are encouraged not to use the latest block
height for their latest announcements/updates but rather to backdate and use
older block heights that they have not used in an announcement/update. There of
course needs to be a limit on the start block height that the node can use:
for `channel_update_2` messages, the first timestamp must be the block height in
which the channel funding transaction was mined and all updates after the
initial one must have increasing timestamps. Nodes are then responsible for
building up their own timestamp buffer: if they want to be able to send
multiple `channel_update_2` messages per block, then they will need to ensure 
that there are blocks during which they do not broadcast any updates. This 
provides an incentive for nodes not to spam the network with too many updates.

To also prevent nodes from building up too large of a burst-buffer with which 
they can spam the network and to give a limit to how low the block height on a 
`node_announcement_2` can be: nodes should not be allowed to use a block height

> smaller 1440 (~ a day worth of blocks) below the current block height.

Reviewer 3: I'm not sure if this spam prevention works in all scenarios:
- A has a block height of 800,144 and sends B a channel_update_2 with height 800,000
- B has a block height of 800,145 and rejects the channel_update_2

Reviewer 3: That's a good point, a buffer of blocks could be used here to avoid this issue

> smaller 1440 (~ a day worth of blocks) below the current block height.

Reviewer 4: ```suggestion
smaller 144 (~ a day worth of blocks) below the current block height.
```

Reviewer 4: I guess the sweet-spot value has to be something low enough to battle spam, but also sufficient for legit gossip messages? Could this depend on the size of the node?

I would expect a big routing node with 100s of channels to produce more gossip than a single-channel node.

## Bootstrapping Taproot Gossip

While the network upgrades from the legacy gossip protocol to the taproot gossip
protocol, the following scenarios may exist for any node on the network:

| scenario | has legacy channels | has taproot channels  | should send `node_announcement` | should send `node_announcement_2` |
|----------|---------------------|-----------------------|---------------------------------|-----------------------------------|
| 1        | no                  | no                    | no                              | no                                |
| 2        | yes                 | no                    | yes                             | no                                |
| 3        | yes                 | yes                   | yes                             | no                                |
| 4        | no                  | yes                   | no                              | yes                               |

### Scenario 1

These nodes have no announced channels and so should not be broadcasting legacy
or new node announcement messages.

### Scenario 2

If a node has legacy channels but no taproot channels, they should broadcast

> only a legacy `node_announcement` message. Both taproot-gossip aware and unaware

Reviewer 2: 👍 

Will be important to make sure things are properly bridged before taproot chans+gossip are deployed in a widespread manner. Nodes may also want to make more "feeler" connections than usual if most of the direct channel peers haven't yet updated as well.

Reviewer 2: Also the various systems deployed for allowing nodes to sync the gossip data over HTTP or w/e will help here as well. Not sure if any of them also allow nodes/users to _post_ gossip data as well though.

nodes, they should continue broadcasting the legacy `node_announcement` 
message.

### Scenario 4

In this scenario, a node has only taproot channels and no legacy channels. This
means that they cannot broadcast a legacy `node_announcement` message since 
un-upgraded nodes will drop this message due to the fact that no legacy 
`channel_announcement` message would have been received for that node. These 
nodes will also not be able to use a legacy `node_announcement` to propagate 
their new `node_announcement_2` message.

### Consideration & Suggestions

While the network is in the upgrade phase, the following suggestions apply:

- Keeping at least a single, announced, legacy channel open during the initial 
  phase of the transition to the new gossip protocol could be very beneficial 
  since nodes can continue to broadcast their legacy `node_announcement` and 
  thus more effectively advertise their support for `option_taproot_gossip`. 
- Nodes are encouraged to actively connect to other nodes that advertise the 
  `option_taproot_gossip` feature bit as this is the only way in which they 
  will learn about taproot channel announcements and updates.

> - Nodes should not use the new `node_announcement_2` message until they have

Reviewer 2: What's the rationale here? If they don't use `node_announcement_2` at all, then none of their taproot channels will propagate.

`option_taproot_gossip` which only makes sense in the context of the old
`node_announcement` message and not the new `node_announcement_v2` message.
The `option_taproot` feature bit can also be implied when using the new taproot 
gossip messages. 

| Bits  | Name                    | Description                                       | Context | Dependencies     |
|-------|-------------------------|---------------------------------------------------|---------|------------------|
| 32/33 | `option_taproot_gossip` | Nodes that understand the taproot gossip messages | IN      | `option_taproot` | 

A node can be assumed to understand the new gossip v2 messages if:

- They advertise `option_taproot_gossip` in a legacy `node_announcement` message
  OR
- They have broadcast one of the new gossip messages defined in this document.

Advertisement of the `option_taproot_gossip` feature bit should be taken to 
mean:

- The node understands and is able to forward all the taproot gossip messages.
- The node may be willing to open an announced taproot channel.

### `open_channel` Extra Requirements

These extra requirements only apply if the `option_taproot` channel type is set

> in the `open_channel` message.

Reviewer 2: IMO channel type becomes even more important in this scenario as one may want to still preferentially open the old segwit v0 channel with certain peers. Assuming the advertised feature bits tell the complete story would _implicltly_ mean that the taproot chans are the new "default", which likely won't be the case for some time (eg: zero fee anchors isn't yet the default for all impls).

> in the `open_channel` message.

Reviewer 5: > Assuming the advertised feature bits tell the complete story would implicltly mean that the taproot chans are the new "default", which likely won't be the case for some time

It's also less future proof. Even if we want TR channels to be assumed default which @Roasbeef already points out is unlikely to be the case for a while, in the future we may have a new default preferred channel type as the network continues to develop and mature.

The sender:

- if `option_taproot_gossip` was negotiated:
    - MAY set the `announce_channel` bit in `channel_flags`
- otherwise:
    - MUST NOT set the `announce_channel` bit.

The receiving node MUST fail the channel if:

- if `option_taproot_gossip` was not negotiated and the `announce_channel`
  bit in the `channel_flags` was set.

### `channel_ready` Extensions

> These extensions only apply if the `option_taproot` channel type was set in the

Reviewer 2: Should specify that these values only need to be attached if an announcement signatures message hasn't already been full constructed. 

We also need to think about the restart implications here as well. Eg: lnd today will sign then stash the `ann_sig` message to read from disk to send to the remote party. However if each time we reconnect a new set of nonces is exchanged, then we'll need to wait for `chan_reest` to be sent (`channel_ready` isn't always sent once a channel is fully operational and has had updates), then we'll need to be ready to regen+ sign again.

`open_channel` message along with the `announce_channel` channel flag.

1. `tlv_stream`: `channel_ready_tlvs`
2. types:

>    1. type: 0 (`announcement_node_pubnonce`)

Reviewer 2: I think it's possible to only send a _single_ (pre aggregated nonce) here, so then the final step would be just combining 2 signatures instead of 4. IIUC, this would require nested musig2 though, which I don't think any one has actually implemented in full.

   2. data:
      * [`66*byte`: `public_nonce`]

>    3. type: 1 (`announcement_bitcoin_pubnonce`)

Author: updated!

  - SHOULD use the MuSig2 `NonceGen` algorithm to generate two unique secret 
    nonces and use these to determine the corresponding public nonces: 
    `announcement_node_pubnonce` and `announcement_bitcoin_pubnonce`.
  
The recipient: 

  - MUST fail the channel if: 
    - if the `announcement_node_pubnonce` or `announcement_bitcoin_pubnonce` tlv 
      fields are missing.
    - if the `announcement_node_pubnonce` and `announcement_bitcoin_pubnonce` 
      are equal.
    
### Rationale

The final signature included in the `channel_announcement_v2` message will be a
single [BIP-340][bip-340] signature that needs to be valid for a public key
derived by aggregating all the `node_ID` and `bitcoin_key` public keys. The
signature for this key will thus be created by aggregating (via MuSig2) four
partial signatures. One for each of the keys: `node_ID_1`, `node_ID_2`, 
`bitcoin_key_1` and `bitcoin_key_2`. A nonce will need to be generated and
exchanged for each of the keys and so each node will need to send the other node
two nonces. The `announcement_node_nonce` is for `node_ID_x` and the 
`announcement_bitcoin_nonce` is for `bitcoin_key_x`.

> Since the channel can only be announced once the `channel_ready` messages have 

Reviewer 2: See comment above: `channel_ready` isn't _always_ sent on reconnect. So if a channel is used before the gossip advertisement is finished, then the nonces won't be available to advertise.

between the two endpoints of a channel and serves as an opt-in mechanism to
allow the announcement of the channel to the rest of the network. 

1. type: xxx (`announcement_signatures_2`)
2. data (tlv_stream):
   1. type: 0 (`channel_id`)
   2. data:
       * [`channel_id`:`channel_id`]
   3. type: 1 (`short_channel_id`)
   4. data:
       * [`short_channel_id`:`short_channel_id`]
   5. type: 2 (`partial_signature`)
   6. data:
       * [`partial_signature`:`partial_signature`]

### Requirements

The requirements are similar to the ones defined for the legacy 
`announcement_signatures`. The below requirements assume that the 
`option_taproot` channel type was set in `open_channel`.

A node:
- if the `open_channel` message has the `announce_channel` bit set AND a 
  `shutdown` message has not been sent:
  - MUST send the `announcement_signatures_2` message.
    - MUST NOT send `announcement_signatures_2` messages until `channel_ready` 
      has been sent and received AND the funding transaction has at least six 
      confirmations.
  - MUST set the `partial_signature` field to the 32-byte `s` value of the
    partial signature calculated as described

>     in [Partial Signature Calculation](#partial-signature-calculation). The

Author: cool I will update it to be more clear. But for now, you can assume that `s`== `partial_sig`

>     in [Partial Signature Calculation](#partial-signature-calculation). The

Author: updated

`announcement_signatures` message, `announcement_signatures_v2` only has one
signature field. This field is a MuSig2 partial signature which is the
aggregation of the two signatures that the sender would have created (one for
`bitcoin_key_x` and another for `node_ID_x`).

## The `channel_announcement_2` Message

This gossip message contains ownership information regarding a taproot channel. 
It ties each on-chain Bitcoin key that makes up the taproot output key to the 
associated Lightning node key, and vice-versa. The channel is not practically 
usable until at least one side has announced its fee levels and expiry, using 
`channel_update_2`.

See [Taproot Channel Proof and Verification](#taproot-channel-proof-and-verification)
for more information regarding the requirements for proving the existence of a
channel.

1. type: xxx (`channel_announcement_2`)
2. data: (tlv_stream):
   1. type: 0 (`announcement_sig`)
   2. data:
        * [`bip340_sig`:`bip340_sig`]
   3. type: 1 (`features`)
   4. data:
        * [`...*byte`: `features`]
   5. type: 2 (`chain_hash`)
   6. data:
        * [`chain_hash`:`chain_hash`]
   7. type: 3 (`short_channel_id`)
   8. data:

>         * [`short_channel_id`:`short_channel_id`]

Author: > IMO we should have an option for a channel to not include an SCID at all

since it is TLV, it can always be made optional in future.

> rather allowing a node to overcommit their balance

I think you are referring to the full on Gossip 2.0 proposal. This proposal is the more vanilla gossip 1.5 that was discussed in oakland.

>         * [`short_channel_id`:`short_channel_id`]

Author: > here's a bunch of complexity to adding the negotiation of building gossip with overcommit, so I wouldn't expect to actually build that now, but there's very little complexity to allowing overcommit now

Could you expand on the details a bit?

>         * [`short_channel_id`:`short_channel_id`]

Reviewer 6: > Then we'd have to define some overcommit ratio (let's roll a dice or bikeshed for months), and a receiver of such a message must track the current amount of bitcoin in utxo-proven channels for each node in their graph, as well as the amount of channels which did not have a utxo proof, and if the utxo-proven channel amount is < amount of non-utxo-proven channels * the overcommit ratio the channels with the lowest (highest? whatever) fake scids are removed.

Should the overcommit amount depend on bitcoin amount in proven channels, or on number of proven channels?

Chain fees scale with number of proven channels, but not necessarily with bitcoin amount in channels.  So I think number of proven channels provides stronger spam protection.

(Sorry if this was discussed in Oakland or elsewhere, I'm new here.)

>         * [`short_channel_id`:`short_channel_id`]

Reviewer 5: > IMO we should have an option for a channel to not include an SCID at all, rather allowing a node to overcommit their balance by simply saying "I already have channel(s) that have utxos, let me announce more without a proof up to multiplier N by not including a proof on this channel".

This is a somewhat unsettling proposition that I'm encountering for the first time. Has this been discussed anywhere as to why people think it's a good idea? I find it unusual that we are suggesting things that resemble leverage/fractional reserve at the protocol level in Bitcoin.

   9. type: 4 (`node_id_1`)
   10. data:
        * [`point`:`point`]
   11. type: 5 (`node_id_2`)
   12. data:
       * [`point`:`point`]
   13. type: 6 (`bitcoin_key_1`)
   14. data:
       * [`point`:`point`]

>    15. type: 7 (`bitcoin_key_2`)

Author: See my response to Rusty's comment above. Basically: I think we can go full-ghost mode once we do Gossip 2.0 (overcommit stuff) but before that (gossip 1.5 aka this) we should have at least _some_ binding to the LN context. The way the proposal is currently designed, nodes can prove that the output cant be spent along the script path. Yes, perhaps we can smush the 2 bitcoin keys into one (i would just need to check that the musig verification can still work)

>    15. type: 7 (`bitcoin_key_2`)

Reviewer 6: > Basically: I think we can go full-ghost mode once we do Gossip 2.0 (overcommit stuff) but before that (gossip 1.5 aka this) we should have at least some binding to the LN context.  The way the proposal is currently designed, nodes can prove that the output cant be spent along the script path.

IIUC, it's important that node1 and node2 prove to *each other* that script path spends are impossible.  But I don't see a reason we need to prove that to the rest of the network.

>    15. type: 7 (`bitcoin_key_2`)

Author: > it's important that node1 and node2 prove to each other that script path spends are impossible. But I don't see a reason we need to prove that to the rest of the network.

i don't think they need to prove this to eachother. They would not have generated the same output key & hence would not have produced the correct signatures if either node added a script path. They would not be able to hide such a path from the peer. 

The reason for the "proving to the network" part is similar to today's reason for advertising the 2 keys: today we prove that the output can only be spent via a 2-of-2 multisig? this makes it a _bit_ more difficult for an attacker to create fake spam since they are now limited in how they can spend the utxo since they would now have to pay more for spending from it (2 sigs). That disincentive is not available with taproot - so the idea here is that we at least show the network that we are limiting how the utxo can be spent in _some_ way - by showing that we cant spend via a script path. But perhaps this is too "cosmetic" to consider sufficient in any case. 

So i think the main thing we should iron out is: do we want to be able to prove (ie, do we care) that the utxo is created via a 2-of-2 musig2 or not. Perhaps a good topic for today's spec call

>    15. type: 7 (`bitcoin_key_2`)

Reviewer 6: > That disincentive is not available with taproot - so the idea here is that we at least show the network that we are limiting how the utxo can be spent in some way - by showing that we cant spend via a script path. But perhaps this is too "cosmetic" to consider sufficient in any case.

The "no-script-path" limitation seems insignificant to me.  I suspect most P2TR outputs today don't have a script path anyway.

> So i think the main thing we should iron out is: do we want to be able to prove (ie, do we care) that the utxo is created via a 2-of-2 musig2 or not.

In the taproot world there's no fee difference between musig and single sig UTXOs.  So spam prevention doesn't seem a compelling reason to prove a UTXO is a musig.

It seems to me that proof-of-UTXO is a "good enough" spam prevention.  In fact, we're talking about moving to overcommit in the future, which would be a weaker spam prevention than is being proposed right here.

>    15. type: 7 (`bitcoin_key_2`)

Author: yep - so I think the next version is gonna make this tweak optional. So by default, nodes will just use the output key found on chain & then if nodes want to include things like the 2 keys used in the musig2 & if they want to add a tweak then they can do so optionally.

   16. data:
       * [`point`:`point`]

### Requirements

The origin node:

- MUST include types 0-7.
- MUST set `chain_hash` to the 32-byte hash that uniquely identifies the chain
  that the channel was opened within:
    - for the _Bitcoin blockchain_:
        - MUST set `chain_hash` value (encoded in hex) equal

>           to `6fe28c0ab6f1b372c1a6a246ae63f74f931e8365e15a089c68d6190000000000`.

Author: [From Rusty](https://github.com/ellemouton/bolts/pull/2#discussion_r1134921767):

I suggest the same trick we use for offers: omitting this field means it is 6fe28c0ab6f1b372c1a6a246ae63f74f931e8365e15a089c68d6190000000000. And you must omit the field if it would be 6fe28c0ab6f1b372c1a6a246ae63f74f931e8365e15a089c68d6190000000000 (remember: both sides need to agree on exactly what this message looks like!).

Note also: I like the use of the TLV here. Because if we want to later, we can adapt this for pre-taproot channels, which would require more / different sigs. Though maybe they'll all close by the time v1-apocolypse is here?

Author: Cool - yeah I like the idea. 
I have heard a few oppositions to assuming defaults though so just gonna leave this thread open to leave room for people to discuss. But personally, I dont see why a default is that bad.

channel was spliced.

## The `node_announcement_2` Message

This gossip message, like the legacy `node_announcement` message, allows a node 
to indicate extra data associated with it, in addition to its public key. 
To avoid trivial denial of service attacks, nodes not associated with an already
known channel (legacy or taproot) are ignored.

Unlike the legacy `node_announcement` message, this message makes use of a 
Schnorr signature instead of an ECDSA one. This will allow nodes to be backed 
by multiple keys since MuSig2 can be used to construct the single signature.

The other two main differences from the legacy message are that the timestamp 
is now a block height and that the message is purely TLV based.

1. type: xxx (`node_announcement_2`)
2. data: (tlv_stream):
    1. type: 0 (`announcement_sig`)
    2. data:
        * [`bip340_sig`:`bip340_sig`]
    3. type: 1 (`features`)
    4. data:
        * [`...*byte`: `features`]
    5. type: 2 (`timestamp`)
    6. data:
        * [`u32`: `block_height`]
    7. type: 3 (`node_ID`)
    8. data: 
        * [`point`:`node_id`]
    9. type: 4 (`color`)
    10. data:

>          * [`rgb_color`:`rgb_color`]

Reviewer 2: Maybe remove the color field? Don't think it was ever seriously used....

On the other hand, if we go TLV everywhere, then it matters less as including it is optional.

    11. type: 5 (`alias`)
    12. data:
         * [`...*utf8`:`alias`]
    13. type: 6 (`ipv4_addrs`)
    14. data: 
         * [`...*ipv4_addr`: `ipv4_addresses`]  
    15. type: 7 (`ipv6_addrs`)
    16. data:
         * [`...*ipv6_addr`: `ipv6_addresses`]
    17. type: 8 (`tor_v3_addrs`)
    18. data: 
         * [`...*tor_v3_addr`: `tor_v3_addresses`]
    19. type: 9 (`dns_hostname`)
    20. data:
         * [`dns_hostname`: `dns_hostname`]

The following subtypes are defined:

1. subtype: `rgb_color`
2. data:
   * [`byte`:`red`]
   * [`byte`:`green`]
   * [`byte`:`blue`]

3. subtype: `ipv4_addr`
4. data:
    * [`u32`:`addr`]
    * [`tu16`:`port`]

5. subtype: `ipv6_addr`
6. data:
    * [`16*byte`:`addr`]
    * [`tu16`:`port`]
 
7. subtype: `tor_v3_addr`
8. data:
    * [`35*utf8`:`onion_addr`]
    * [`tu16`:`port`]

9. subtype: `dns_hostname`
10. data:
     * [`...*utf8`:`hostname`]

>      * [`tu16`:`port`]

Author: [From Rusty](https://github.com/ellemouton/bolts/pull/2#discussion_r1134932702):

And if port is 0, say it's 9735?

Author: same here as for the above re defaults

messages: nodes are allowed to send at most one announcement and update per
block. To allow for bursts, nodes are encouraged not to use the latest block
height for their latest announcements/updates but rather to backdate and use
older block heights that they have not used in an announcement/update. There of
course needs to be a limit on the start block height that the node can use:
for `channel_update_2` messages, the first timestamp must be the block height in
which the channel funding transaction was mined and all updates after the
initial one must have increasing timestamps. Nodes are then responsible for
building up their own timestamp buffer: if they want to be able to send
multiple `channel_update_2` messages per block, then they will need to ensure 
that there are blocks during which they do not broadcast any updates. This 
provides an incentive for nodes not to spam the network with too many updates.

To also prevent nodes from building up too large of a burst-buffer with which 
they can spam the network and to give a limit to how low the block height on a 
`node_announcement_2` can be: nodes should not be allowed to use a block height

> smaller 2016 (~ one week worth of blocks) below the current block height.

Reviewer 5: ```suggestion
smaller 2016 (~ two weeks worth of blocks) below the current block height.
```

Author: i've removed this part :)

define a new feature bit, `option_taproot_gossip` to be included in the `init`
message as well as the _old_ `node_announcement` message. Note that for all
other feature bits with the `N` and `C` contexts, it can be assumed that those
contexts will switch to the new `node_announcement_2` and
`channel_announcement_2` messages for all feature bits _except_ for
`option_taproot_gossip` which only makes sense in the context of the old
`node_announcement` message and not the new `node_announcement_v2` message.
The `option_taproot` feature bit can also be implied when using the new taproot 
gossip messages. 

| Bits  | Name                    | Description                                       | Context | Dependencies     |
|-------|-------------------------|---------------------------------------------------|---------|------------------|
| 32/33 | `option_taproot_gossip` | Nodes that understand the taproot gossip messages | IN      | `option_taproot` | 

A node can be assumed to understand the new gossip v2 messages if:

- They advertise `option_taproot_gossip` in the `init` or 
  legacy `node_announcement` messages
  OR

> - They have broadcast one of the new gossip messages defined in this document.

Reviewer 5: I think it's probably a good idea to require consistency here. I don't love the idea of a node broadcasting Gossip v2 messages without having advertised `option_taproot_gossip` at all. I'm open to being challenged here. I think it's fair to say that if they broadcast it, they do understand it, as you note. However, I think things get significantly more complicated on the implementation side if we have to infer the node's ability to understand from either of these somewhat independent informational sources.

Author: agreed - I've removed this 👍

   2. data:
      * [`66*byte`: `public_nonce`]
   3. type: 2 (`announcement_bitcoin_pubnonce`)
   4. data:
      * [`66*byte`: `public_nonce`]
   
### Requirements

The sender: 

  - MUST set the `announcement_node_pubnonce` and 
    `announcement_bitcoin_pubnonce` tlv fields in `channel_ready`.
  - SHOULD use the MuSig2 `NonceGen` algorithm to generate two unique secret 
    nonces and use these to determine the corresponding public nonces: 
    `announcement_node_pubnonce` and `announcement_bitcoin_pubnonce`.
  
The recipient: 

  - MUST fail the channel if: 
    - if the `announcement_node_pubnonce` or `announcement_bitcoin_pubnonce` tlv 
      fields are missing.
    - if the `announcement_node_pubnonce` and `announcement_bitcoin_pubnonce` 
      are equal.
    
### Rationale

The final signature included in the `channel_announcement_v2` message will be a
single [BIP-340][bip-340] signature that needs to be valid for a public key
derived by aggregating all the `node_ID` and `bitcoin_key` public keys. The
signature for this key will thus be created by aggregating (via MuSig2) four
partial signatures. One for each of the keys: `node_ID_1`, `node_ID_2`, 

> `bitcoin_key_1` and `bitcoin_key_2`. A nonce will need to be generated and

Reviewer 5: ```suggestion
`bitcoin_key_1` and `bitcoin_key_2`. A (MuSig) nonce will need to be generated and
```

exchanged for each of the keys and so each node will need to send the other node
two nonces. The `announcement_node_nonce` is for `node_ID_x` and the 
`announcement_bitcoin_nonce` is for `bitcoin_key_x`.

Since the channel can only be announced once the `channel_ready` messages have 
been exchanged and since it is generally preferred to keep nonce exchanges as 

> Just In Time as possible, the nonces are exchanged via the `channel_ready` 

Reviewer 5: Can we include a citation for the Just In Time preference? Reading [BIP327](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#design), it seems to suggest in bullet 3 that it doesn't really matter how early you exchange them since it is valid to do so even before the exact signing participants have been determined.

   3. type: 2 (`short_channel_id`)
   4. data:
       * [`short_channel_id`:`short_channel_id`]
   5. type: 4 (`partial_signature`)
   6. data:
       * [`partial_signature`:`partial_signature`]

### Requirements

The requirements are similar to the ones defined for the legacy 
`announcement_signatures`. The below requirements assume that the 
`option_taproot` channel type was set in `open_channel`.

A node:
- if the `open_channel` message has the `announce_channel` bit set AND a 
  `shutdown` message has not been sent:
  - MUST send the `announcement_signatures_2` message once `channel_ready`
      has been sent and received AND the funding transaction has at least six 
      confirmations.
  - MUST set the `partial_signature` field to the 32-byte `partial_sig` value of
    the partial signature calculated as described in [Partial Signature 
    Calculation](#partial-signature-calculation). The message to be signed is
    `MsgHash("channel_announcement", "announcement_sig", m)` where `m` is the
    serialisation of the `channel_announcement_2` message excluding the
    `announcement_sig` field (see the
    [`MsgHash`](#signature-message-construction) definition).
- otherwise:

>     - MUST NOT send the `announcement_signatures_2` message.

Author: good point. 

I took the original logic from the legacy `announcement_signatures`. But yeah - perhaps it is possible to construct it sooner as long as it isnt broadcast before 6 confs

A recipient node:
- if the `short_channel_id` is NOT correct:
    - SHOULD send a `warning` and close the connection, or send an
      `error` and fail the channel.
- if the `partial_signature` is NOT valid as
  per [Partial Signature Verification](#partial-signature-verification):
    - MAY send a `warning` and close the connection, or send an
      `error` and fail the channel.
- if it has sent AND received a valid `announcement_signatures_2` message:
    - SHOULD queue the `channel_announcement_2` message for its peers.
- if it has not sent `channel_ready`:
    - MAY send a `warning` and close the connection, or send an `error` and fail
      the channel.

### Rationale

The message contains the necessary partial signature, by the sender, that the
recipient will be able to combine with their own partial signature to construct
the signature to put in the `channel_announcement_v2` message. Unlike the legacy
`announcement_signatures` message, `announcement_signatures_v2` only has one
signature field. This field is a MuSig2 partial signature which is the
aggregation of the two signatures that the sender would have created (one for

> `bitcoin_key_x` and another for `node_ID_x`).

Reviewer 5: 👍🏼

`channel_update_2`.

See [Taproot Channel Proof and Verification](#taproot-channel-proof-and-verification)
for more information regarding the requirements for proving the existence of a
channel.

1. type: xxx (`channel_announcement_2`)
2. data: (tlv_stream):
   1. type: 0 (`announcement_sig`)
   2. data:
        * [`bip340_sig`:`bip340_sig`]
   3. type: 1 (`chain_hash`)
   4. data:
       * [`chain_hash`:`chain_hash`]
   5. type: 2 (`features`)
   6. data:
        * [`...*byte`: `features`]
   7. type: 4 (`short_channel_id`)
   8. data:
        * [`short_channel_id`:`short_channel_id`]
   9. type: 6 (`node_id_1`)
   10. data:
        * [`point`:`point`]
   11. type: 8 (`node_id_2`)
   12. data:
       * [`point`:`point`]
   13. type: 3 (`bitcoin_key_1`)
   14. data:
       * [`point`:`point`]
   15. type: 5 (`bitcoin_key_2`)
   16. data:
       * [`point`:`point`]
   17. type: 7 (`tap_tweak`)
   18. data:
       * [`point`:`point`]

### Requirements

The origin node:

- If the chain being referred to is not the Bitcoin blockchain:
  - MUST set `chain_hash` to the 32-byte hash that uniquely identifies the chain
    that the channel was opened within:

> - otherwise, MUST not set `chain_has` as the _Bitcoin blockchain_ is assumed. 

Reviewer 5: ```suggestion
- otherwise, MUST not set `chain_hash` as the _Bitcoin blockchain_ is assumed. 
```

      in [Taproot Channel Proof and Verification](#taproot-channel-proof-and-verification).
- MUST set `node_id_1` and `node_id_2` to the public keys of the two nodes
  operating the channel, such that `node_id_1` is the lexicographically-lesser
  of the two compressed keys sorted in ascending lexicographic order.
- MUST set `bitcoin_key_1` and `bitcoin_key_2` to `node_id_1` and `node_id_2`'s
  respective `funding_pubkey`s.
- MUST set `features` based on what features were negotiated for this channel,
  according to [BOLT #9](09-features.md#assigned-features-flags)
- MUST set `announcement_sig` signature to the [BIP-340][bip-340] signature
  calculated by passing the partial signatures sent and received during the
  `announcement_signatures_2` exchange to the
  [MuSig2 PartialSigAgg][musig-partial-sig-agg] function.

The receiving node:

- If any even typed messages are not present:
    - SHOULD send a `warning`.
    - MAY close the connection.
    - MUST ignore the message.
- MUST verify the integrity AND authenticity of the message by verifying the
  `announcement_sig` as per [BIP 340][bip-340-verify].
- if there is an unknown even bit in the `features` field:

>     - MUST NOT attempt to route messages through the channel.

Reviewer 5: what is meant by "messages" here? Does it go beyond HTLCs?

advertisements in their database to cater for gossip syncing with older
peers, but they should favour the new protocol when making routing decisions.

While the network is in the upgrade phase, nodes will likely want to advertise
on both the old and new protocols for P2WSH channels so that older nodes
continue to see their channels. Eventually, when most of the network has
advertised their understanding of the new protocol, nodes can stop advertising
on the old protocol.

## Specification

### Type Definitions

The following convenient types are defined:

* `bip340sig`: a 64-byte bitcoin Elliptic Curve Schnorr signature as
  per [BIP-340][bip-340].
* `partial_signature`: a 32-byte partial MuSig2 signature as defined
  in [BIP-MuSig2][bip-musig2].
* `public_nonce`: a 66-byte public nonce as defined in [BIP-MuSig2][bip-musig2].
* `utf8`: a byte as part of a UTF-8 string. A writer MUST ensure an array of
  these is a valid UTF-8 string, a reader MAY reject any messages containing an
  array of these which is not a valid UTF-8 string.

### Pure TLV messages

All the messages defined in this document are pure TLV streams. The signed TLV

> range is defined as the inclusive ranges: 0 to 159 and 1000000000 to 2999999999.

Reviewer 7: This will be probably clarified with test vectors, but just to make sure that I'm understanding this correctly, what we're signing is just the encoded TLV stream (with the normal TLV stream rules of ordering by TLV tag) without a length prefix and without the message type (because it will be included in the use of `MsgHash`)?

For example for a node announcement, we would just sign:

```
(0x00 <features_len> <feature_bits>) (0x02 0x04 <block_height>) ...
```

Or for a channel announcement, we would sign:

```
(0x00 0x20 <block_hash>) (0x02 <features_len> <feature_bits>) (0x04 0x08 <short_channel_id>) ...
```

Reviewer: Also, this has changed in bolt12 (a long time ago!), so signatures start at 240.  Let's update this too, and move signature fields to 240?

And I assume the same as @t-bast : we're signing the tlv, not the message.

indicates that a node is able to produce an SPV proof and attach it to its
`channel_announcement_2` message if asked.

### `open_channel` Extensions

- If the `option_taproot` channel type was set in `open_channel`:
    - The sender:
        - If `option_gossip_v2` was negotiated:
            - MAY set the `announce_channel` bit in `channel_flags`
        - otherwise:
            - MUST NOT set the `announce_channel` bit.

    - The receiver:
        - if `option_gossip_v2` was not negotiated and the `announce_channel` bit
          in `channel_flags` was set, MUST fail the channel.

#### Rationale

- The `option_taproot` channel type was defined before this spec and limited
  peers that negotiated the channel type to open unannounced channels only.
  This document defines how peers can go about announcing these channels, and so
  this restriction can be lifted.

### `channel_ready` Extensions

The following extensions are defined for the `channel_ready` message:

1. `tlv_stream`: `channel_ready_tlvs`
2. types:
    1. type: 0 (`announcement_node_pubnonce`)
    2. data:
        * [`66*byte`: `public_nonce`]
    3. type: 2 (`announcement_bitcoin_pubnonce`)
    4. data:

>         * [`66*byte`: `public_nonce`]

Reviewer 7: Can you also add those TLVs to `splice_locked`? This will be necessary to announce splices, exactly like what is done for the initial channel funding.

#### Requirements:

The following requirements apply if the `announce_channel` bit was set in the
`channel_flags` field of the `open_channel` message. Nodes advertising the
`option_gossip_announce_private` bit, may also choose to send the
`channel_ready` message at any point during the channel's lifetime in which
case, the following requirements also apply even if the `announce_channel` bit
was not set in `open_channel`.

The sender:

- If the negotiated channel type is `option_taproot` and `option_gossip_v2` was
  negotiated, then when the node is ready to advertise the channel:
    - SHOULD send `channel_ready` with the `announcement_node_pubnonce` and
      `announcement_bitcoin_pubnonce` fields set. The `announcement_node_pubnonce`
      must be set to the public nonce to be used for signing with the node's
      bitcoin key and the `announcement_bitcoin_pubnonce` must be set to the
      public nonce to be used for signing with the node's node ID key.
    - Upon reconnection, if a fully signed `channel_announcement_2` has not yet
      been constructed:

>         - SHOULD re-send `channel_ready` with the nonce fields set.

Reviewer 7: Re-sending `channel_ready` on reconnection to finalize announcements won't be compatible with splicing. We introduced a `my_current_funding_locked` TLV (see https://github.com/lightning/bolts/blob/master/02-peer-protocol.md#message-retransmission) that contains a `retransmit_flags` bitfield where bit `0` can ask for a retransmission of `announcement_signatures`.

We should use bit `1` of those `retransmit_flags` to request a retransmission of `announcement_signatures_2` and introduce a new `channel_reestablish` TLV that contains announcement nonces for the transaction matching `my_current_funding_locked`.

```
1. type: 7 (`announcement_nonces`)
2. data:
    * [`sha256`:`funding_txid`]
    * [`66*byte`: `node_public_nonce`]
    * [`66*byte`: `bitcoin_public_nonce`]
```

With this, on reconnection, we don't need any `channel_ready` retransmission. The requirements should be:

```
- Both nodes MUST include the `announcement_nonces` TLV with `funding_txid` matching their `my_current_funding_locked` (even if they already published a channel_announcement_2, because their peer may not have been able to do the same if announcement_signatures_2 was lost during a disconnection)
- If nodes haven't yet created channel_announcement_2, they MUST set bit 1 in my_current_funding_locked.retransmit_flags
- When receiving my_current_funding_locked.retransmit_flags with bit 1 set, nodes MUST send a matching announcement_signatures_2
```

I'm not sure if we should include the `funding_txid` or just specify that it applies for the value in `my_current_funding_locked`.

### `announcement_signatures` Extra requirements

If peers have negotiated the `option_gossip_announce_private` bit, then the
`announcement_signatures` message can be used to initiate the signing flow for
a previously unannounced channel via the `channel_announcement` message at
any point during the channel's lifecycle. All existing requirements of the
message remain.

### The `announcement_signatures_2` Message

Like the legacy `announcement_signatures` message, this is a direct message
between the two endpoints of a channel and serves as an opt-in mechanism to
allow the announcement of the channel to the rest of the network. This message
is used to exchange the partial signatures required to construct the final
signature for the `channel_announcement_2` message.

This message can be sent in two cases:

1) At the start of the channel's lifetime if the `announce_channel` bit was set
   in the `channel_flags` field of the `open_channel` message.
2) At any other point in the channel's lifetime if the
   `option_gossip_announce_private` bit was negotiated between the channel peers.

1. type: 260 (`announcement_signatures_2`)
2. data:
    * [`announcement_signatures_2_tlvs`:`tlvs`]

> 1. `tlv_stream`: `announcement_signatures_2_tlvs`

Reviewer 7: Can you also add a TLV containing the funding txid which this announcement is for (necessary for splicing)?

```
1. type: 6 (`funding_txid`)
2. data:
    * [`sha256`:`funding_txid`] 
```

For the current `announcement_signatures` message we use the `scid` to match the correct funding transaction, but it would be cleaner to use the `funding_txid` explicitly.

negotiated.

A node:
- if the `open_channel` message has the `announce_channel` bit set AND a
  `shutdown` message has not been sent:
    - MUST send the `announcement_signatures_2` message once a `channel_ready`
      message containing the announcement nonces has been sent and received AND
      the funding transaction has at least six confirmations.
    - MUST set the `partial_signature` field to the 32-byte `partial_sig` value
      of the partial signature calculated as described in [Partial Signature
      Calculation](#partial-signature-calculation). The message to be signed is
      `MsgHash("channel_announcement", "signature", m)` where `m` is the
      serialisation of the `channel_announcement_2` message tlv stream (see the
      [`MsgHash`](#signature-message-construction) definition).
- otherwise if the `option_gossip_announce_private` bit has been negotiated:
    - MAY send the `announcement_signatures_2` message at any point during the
      channel's lifetime with the same constraints as defined above.
- otherwise:
    - MUST NOT send the `announcement_signatures_2` message.
- upon reconnection (once the above timing requirements have been met):

>     - MUST re-send a `channel_ready` message with the nonce fields set and await

Reviewer 7: As highlighted in my other comment, we shouldn't retransmit `channel_ready` on reconnection for this, but rather rely on `channel_reestablish` TLVs (which also guarantees atomicity, which makes the reconnection state machine simpler).

recipient will be able to combine with their own partial signature to construct
the signature to put in the `channel_announcement_2` message. Unlike the legacy
`announcement_signatures` message, `announcement_signatures_2` only has one
signature field. This field is a MuSig2 partial signature which is the
aggregation of the two signatures that the sender would have created (one for
`bitcoin_key_x` and another for `node_ID_x`).

### The `channel_announcement_2` Message

This gossip message contains ownership information regarding either a P2TR or
P2WSH channel. The channel is not practically usable until at least one side has
announced its fee levels and expiry, using `channel_update_2`.

1. type: 267 (`channel_announcement_2`)
2. data:
    * [`channel_announcement_2_tlvs`:`tlvs`]

1. `tlv_stream`: `channel_announcement_2_tlvs`
2. types:
    1. type: 0 (`chain_hash`)
    2. data:
        * [`chain_hash`:`chain_hash`]
    1. type: 2 (`features`)
    2. data:
        * [`...*byte`:`features`]
    1. type: 4 (`short_channel_id`)
    2. data:
        * [`short_channel_id`:`short_channel_id`]
    1. type: 6 (`capacity_satoshis`)
    2. data:

>         * [`u64`:`capacity_satoshis`]

Reviewer: tu64?

| 0  (`chain_hash`)                  | `6fe28c0ab6f1b372c1a6a246ae63f74f931e8365e15a089c68d6190000000000` | The hash of the genesis block of the mainnet Bitcoin blockchain.                            | 

### The `node_announcement_2` Message

This gossip message, like the `node_announcement` message, allows a node
to indicate extra data associated with it, in addition to its public key.
To avoid trivial denial of service attacks, nodes not associated with an already
known channel (advertised via a `channel_announcement_2` message) are ignored.

Unlike the `node_announcement` message, this message makes use of a
BIP340 signature instead of an ECDSA one. This will allow nodes to be backed
by multiple keys since MuSig2 can be used to construct the single signature.

1. type: 269 (`node_announcement_2`)
2. data:
    * [`node_announcement_2_tlvs`:`tlvs`]

1. `tlv_stream`: `node_announcement_2_tlvs`
2. types:
    1. type: 0 (`features`)
    1. type: 1 (`color`)
    2. data:
        * [`rgb_color`:`rgb_color`]
    2. data:
        * [`...*byte`: `features`]
    1. type: 2 (`block_height`)
    1. type: 3 (`alias`)
    2. data:
        * [`...*utf8`:`alias`]
    2. data:

>         * [`u32`: `block_height`]

Reviewer: This got really confused somehow, with fields all mixed together, and tools/extract_formats.py didn't like it (also, due to extra spacing between : and `. 

I made this patch:

```diff
diff --git a/bolt-taproot-gossip.md b/bolt-taproot-gossip.md
index d180297..f6f9f8d 100644
--- a/bolt-taproot-gossip.md
+++ b/bolt-taproot-gossip.md
@@ -658,32 +658,32 @@ by multiple keys since MuSig2 can be used to construct the single signature.
 1. `tlv_stream`: `node_announcement_2_tlvs`
 2. types:
     1. type: 0 (`features`)
+    2. data:
+        * [`...*byte`:`features`]
     1. type: 1 (`color`)
     2. data:
         * [`rgb_color`:`rgb_color`]
-    2. data:
-        * [`...*byte`: `features`]
     1. type: 2 (`block_height`)
+    2. data:
+        * [`u32`:`block_height`]
     1. type: 3 (`alias`)
     2. data:
         * [`...*utf8`:`alias`]
-    2. data:
-        * [`u32`: `block_height`]
     1. type: 4 (`node_id`)
     2. data:
         * [`point`:`node_id`]
     1. type: 5 (`ipv4_addrs`)
     2. data:
-        * [`...*ipv4_addr`: `ipv4_addresses`]
+        * [`...*ipv4_addr`:`ipv4_addresses`]
     1. type: 7 (`ipv6_addrs`)
     2. data:
-        * [`...*ipv6_addr`: `ipv6_addresses`]
+        * [`...*ipv6_addr`:`ipv6_addresses`]
     1. type: 9 (`tor_v3_addrs`)
     2. data:
-        * [`...*tor_v3_addr`: `tor_v3_addresses`]
+        * [`...*tor_v3_addr`:`tor_v3_addresses`]
     1. type: 11 (`dns_hostnames`)
     2. data:
-        * [`...*dns_hostname`: `dns_hostnames`]
+        * [`...*dns_hostname`:`dns_hostnames`]
     1. type: 160 (`signature`)
     2. data:
         * [`bip340sig`:`sig`]
```

bits here refer to the nodes, not the `node_announcement_2` message itself.

### Security Considerations for Node Aliases

The security considerations for node aliases mentioned in
[BOLT #7][bolt-7-alias-security] apply here too.

### The `channel_update_2` Message

After a channel has been initially announced via `channel_announcement_2`, each
side independently announces the fees and minimum expiry delta it requires to
relay HTLCs through this channel. Each uses the 8-byte short channel id that
matches the `channel_announcement_2` and the `second_peer` field to
indicate which end of the channel it's on (origin or final). A node can do this
multiple times, in order to change fees.

1. type: 271 (`channel_update_2`)
2. data:
    * [`channel_update_2_tlvs`:`tlvs`]

1. `tlv_stream`: `channel_update_2_tlvs`
2. types:
    1. type: 0 (`chain_hash`)
    2. data:
        * [`chain_hash`:`chain_hash`]
    1. type: 2 (`short_channel_id`)
    2. data:
        * [`short_channel_id`:`short_channel_id`]
    1. type: 4 (`block_height`)
    2. data:
        * [`u32`:`block_height`]
    1. type: 6 (`disable_flags`)
    2. data:
        * [`byte`:`disable_flags`]

>     1. type: 8 (`second_peer`)

Reviewer: Can we please reuse the sciddir_or_pubkey type in the short_channel_id field, except insist it be an scidd (not a pubkey)?  THe format in bolt 1 is <dirbyte><scid>?

    1. type: 10 (`cltv_expiry_delta`)
    2. data:
        * [`u16`:`cltv_expiry_delta`]
    1. type: 12 (`htlc_minimum_msat`)
    2. data:
        * [`u64`:`htlc_minimum_msat`]
    1. type: 14 (`htlc_maximum_msat`)
    2. data:
        * [`u64`:`htlc_maximum_msat`]

>     1. type: 16 (`fee_base_msat`)

Reviewer: I suggest tu types here, to save some space.

>     2. data:

Author: yes indeed- inbound fees will be included here 👍

|------------------------------------|--------------------------------------------------------------------|---------------------------------------------------------------------------------------------|
| 0  (`chain_hash`)                  | `6fe28c0ab6f1b372c1a6a246ae63f74f931e8365e15a089c68d6190000000000` | The hash of the genesis block of the mainnet Bitcoin blockchain.                            | 
| 6  (`disable`)                     | empty                                                              |                                                                                             | 
| 10 (`cltv_expiry_delta`)           | 80                                                                 |                                                                                             | 
| 12 (`htlc_minimum_msat`)           | 1                                                                  |                                                                                             | 

> | 14 (`htlc_maximum_msat`)           | floor(channel capacity / 2)                                        | // TODO: remove this since makes encoding/decoding dependent on things outside the message? | 

Reviewer: I don't mind this, it's just arbitrary.  We would default to 99%, which is reasonable, but 100% would work too.

The receiving node:

- If the `short_channel_id` does NOT match a previous `channel_announcement_2`
  or `channel_announcement`, OR if the channel has been closed in the meantime:
    - MUST ignore `channel_update_2`s that do NOT correspond to one of its own
      channels.
- SHOULD accept `channel_update_2`s for its own channels (even if non-public),
  in order to learn the associated origin nodes' forwarding parameters.
- if `signature` is NOT a valid [BIP340][bip-340] signature (using `node_id`
  over the message):
    - SHOULD send a `warning` and close the connection.
    - MUST NOT process the message further.

### Query Messages

TODO:
1. new first block height & block range fields in gossip_timestamp_range
2. add block-height query option (like timestamps query option)

# Appendix A: Algorithms

## Partial Signature Calculation

When both nodes have exchanged a `channel_ready` message containing the
`announcement_node_pubnonce` and `announcement_bitcoin_pubnonce` fields then
they will each have the following information:

- `node_1` will know:
    - `bitcoin_priv_key_1`, `node_ID_priv_key_1`,
    - `bitcoin_key_2`,

>     - `node_ID_1`,

Reviewer 8: Typo: should this be node_ID_2?

I mean, node_ID_1 refers to the local node itself, so it is already known. node_ID_2, on the other hand, is learned during the handshake protocol.

The same comment applies below for node_ID_2.

       ```
       partial_sig_node_2 = MuSig2.Sign(announcement_node_secnonce_2, node_ID_priv_key_2, session_ctx) 
       ```
    - calculates a partial signature for `bitcoin_ID_2` as follows:
       ```
       partial_sig_bitcoin_2 = MuSig2.Sign(announcement_bitcoin_secnonce_2, bitcoin_priv_key_2, session_ctx) 
       ```
    - calculates `partial_sig_2` as follows:
       ```
       partial_sig_2 =  MuSig2.PartialSigAgg(partial_sig_node_2, partial_sig_bitcoin_2, session_ctx)
       ```     

Note that since there are no tweaks involved in this MuSig2 signing flow,
signature aggregation is simply the addition of the two signatures:

```
    partial_sig = (partial_sig_node + partial_sig_bitcoin) % n
```

Where `n` is the [secp256k1][secp256k1] curve order.

## Partial Signature Verification

Since the partial signature put in `announcement_signatures_2` is the addition
of two of four signatures required to make up the final MuSig2 signature, the
verification of the partial signature is slightly different from what is
specified in the MuSig2 spec. The slightly adjusted algorithm will be defined
here. The notation used is the same as defined in the [MuSig2][musig-notation]

> spec.

Reviewer 7: Why do we want to modify musig2? This adds the extra cost of implementing and maintaining a different set of musig2 interfaces from the standard ones just for this, which IMO is a very bad idea. 

Can't we just include two partial sigs in `announcement_signatures_2` (one with the bitcoin key and one with the lightning key, which can be bundled in a single 64-bytes TLV) and just use the standard musig2 4-of-4 algorithm? I'd really like to use vanilla musig2 interfaces as much as possible, if this was only done to save a few bytes on the `announcement_signatures_2` message I think this is overkill (it's a small message that isn't exchanged often, clearly not where optimization is useful) and risky.

Reviewer: Yes please.  I haven't implemented it yet, though, so I don't know *how much* pain I'm going to be in...

[open-chan-msg]: ./02-peer-protocol.md#the-open_channel-message
[ml-rusty-2019-gossip-v2]: https://lists.linuxfoundation.org/pipermail/lightning-dev/2019-July/002065.html
[ml-roasbeef-2022-tr-chan-announcement]: https://lists.linuxfoundation.org/pipermail/lightning-dev/2022-March/003526.html
[bip-340]: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki
[bip-340-verify]: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#verification
[simple-taproot-chans]: https://github.com/lightning/bolts/pull/995
[musig-keysort]: https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki#key-sorting
[musig-keyagg]: https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki#key-aggregation
[musig-signing]: https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki#signing
[bip86-tweak]: https://github.com/bitcoin/bips/blob/master/bip-0086.mediawiki#address-derivation
[bip-musig2]: https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki
[musig-session-ctx]: https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki#session-context
[musig-notation]: https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki#notation

> [musig-partial-sig-agg]: https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki#partial-signature-aggregation

Reviewer 8: Several references appear to point to outdated Jonas Nick draft locations and currently resolve to broken links. Should these reference the finalized BIP327 document instead?

a `channel_announcement_2`. This makes it easier to reason about the protocol.
Any nodes that understand both protocols are encouraged to persist both
advertisements in their database to cater for gossip syncing with older
peers, but they should favour the new protocol when making routing decisions.

While the network is in the upgrade phase, nodes will likely want to advertise
on both the old and new protocols for P2WSH channels so that older nodes
continue to see their channels. Eventually, when most of the network has
advertised their understanding of the new protocol, nodes can stop advertising
on the old protocol.

## Specification

### Type Definitions

The following convenient types are defined:

* `bip340sig`: a 64-byte bitcoin Elliptic Curve Schnorr signature as
  per [BIP-340][bip-340].
* `partial_signature`: a 32-byte partial MuSig2 signature as defined
  in [BIP-MuSig2][bip-musig2].
* `public_nonce`: a 66-byte public nonce as defined in [BIP-MuSig2][bip-musig2].
* `utf8`: a byte as part of a UTF-8 string. A writer MUST ensure an array of
  these is a valid UTF-8 string, a reader MAY reject any messages containing an

>   array of these which is not a valid UTF-8 string.

Reviewer 9: do `bip340sig` and `utf8` need to be specified here given that they are already in bolt-01?

        - if `merkle_root_hash` is present:
            - validation fails.
        - if the last 32 bytes of the scriptpubkey do not match the SHA256 of `2 <lesser_bitcoin_key> <greater_bitcoin_key> 2 OP_CHECKMULTISIG` (where lesser and greater are the `bitcoin_key_1` and `bitcoin_key_2` ordered as defined in [BOLT 3](03-transactions.md#funding-transaction-output)):
            - validation fails
    - otherwise (P2TR):
        - Let P_internal = KeyAgg(MuSig2.KeySort(`bitcoin_key_1`, `bitcoin_key_2`))
        - if `merkle_root_hash` is not present:
            - if `taproot_key` != P_internal:
                - validation fails
        - otherwise:
            - let `p` = `bytes(P_internal)`
            - let `t = H("TapTweak", p || merkle_root_hash)` where `H` is the BIP340 tagged hash function defined in [BOLT #12][bolt-12]
            - if `taproot_key != P_internal + t*G`:
                - validation fails
    - Let P_agg = MuSig2.KeyAgg(MuSig2.KeySort(`node_id_1`, `node_id_2`, `bitcoin_key_1`, `bitcoin_key_2`))
- otherwise (not both bitcoin_key fields):
    - if either `bitcoin_key_1` or `bitcoin_key_2` are present:
        - validation fails

>     - Let P_agg = MuSig2.KeyAgg(MuSig2.KeySort(`node_id_1`, `node_id_2`, taproot_key))

Reviewer 9: I may be missing context but I do not understand why allow the option of specifying `bitcoin_key_1/2` but no `merkle_root_hash` in the case of a `P2TR`? For how simple taproot channels are constructed, the `merkle_root_hash` will need to be present and if this is some construct outside of simple taproot channels, then the bitcoin keys can be omitted and use the public key in the taproot output directly.

- `cltv_expiry_delta` is the number of blocks that the node will subtract from
  an incoming HTLC's `cltv_expiry`.
- `htlc_minimum_msat` is the minimum HTLC value (in millisatoshi) that the
  channel peer will accept.
- `htlc_maximum_msat` is the maximum value the node will send through this
  channel for a single HTLC.
    - MUST be less than or equal to the channel capacity
    - MUST be less than or equal to the `max_htlc_value_in_flight_msat` it
      received from the peer in the `open_channel` message.
- `fee_base_msat` is the **outbound** base fee (in millisatoshis) that the node
  will charge for forwarding an HTLC out via this channel.
- `fee_proportional_millionths` is the **outbound** amount (in millionths of a
  satoshi) that the node will charge per transferred satoshi for forwarding an
  HTLC out via this channel.
- `inbound_fee_base_msat` is the additional base fee (in millisatoshis) that
  the node charges for forwarding an HTLC that *arrived* on this channel,
  regardless of which channel it is forwarded out via. This fee is positive
  only (`tu32`), meaning a node MAY surcharge for receiving HTLCs over the

>   channel but MUST NOT advertise a discount.

Reviewer 10: I'm not sure "MUST NOT advertise a discount" makes sense here - its not just MUST NOT its "can not" because there is simply no way to encode it at all.

      the channel specified in the `channel_announcement_2` message.
- MUST set `signature` to a valid [BIP340][bip-340] signature for its own
  `node_id` key. The message to be signed is
  `MsgHash("channel_update_2", "signature", m)` where `m` is the serialised
  TLV stream of the `channel_update` (see the
  [`MsgHash`](#signature-message-construction) definition).
- SHOULD NOT create redundant `channel_update_2`s.
- If it creates a new `channel_update_2` with updated channel parameters:
    - SHOULD keep accepting the previous channel parameters for 2 blocks.

The receiving node:

- If `short_channel_id`, `block_height` or `signature` fields are missing:
    - SHOULD send a `warning`.
    - MAY close the connection.
    - MUST ignore the message.
- If `short_channel_id` is not in the `sciddir` form (i.e. its first byte is
  neither `0` nor `1`):
    - SHOULD send a `warning`.
    - MAY close the connection.
    - MUST ignore the message.
- Let `signing_node_id` be `node_id_1` if the `short_channel_id` direction
  byte is `0`, and `node_id_2` otherwise.
- If the channel-id portion of `short_channel_id` does NOT match a previous

>   `channel_announcement_2` or `channel_announcement`, OR if the channel has

Reviewer 9: I don't think this should have `channel_announcement` since as mentioned above they are to be treated separate.

  `signing_node_id` over `MsgHash("channel_update_2", "signature", m)`):
    - SHOULD send a `warning` and close the connection.
    - MUST ignore the message.
- if `block_height` is less than
  `current_block_height − max_backdate_blocks` (see
  [Rate Limiting](#rate-limiting)):
    - SHOULD ignore the message.

### Query Messages

TODO:
1. new first block height & block range fields in gossip_timestamp_range
2. add block-height query option (like timestamps query option)

# Appendix A: Algorithms

## Partial Signature Calculation

When both nodes have exchanged a `channel_ready` message containing the
`announcement_node_pubnonce` and `announcement_bitcoin_pubnonce` fields then
they will each have the following information:

- `node_1` will know:
    - `bitcoin_priv_key_1`, `node_ID_priv_key_1`,
    - `bitcoin_key_2`,
    - `node_ID_1`,
    - `announcement_node_secnonce_1` and `announcement_node_pubnonce_1`,
    - `announcement_bitcoin_secnonce_1` and `announcement_bitcoin_pubnonce_1`,
    - `announcement_node_pubnonce_2`,
    - `announcement_bitcoin_pubnonce_2`,

- `node_2` will know:
    - `bitcoin_priv_key_2`, `node_ID_priv_key_2`,
    - `bitcoin_key_1`,

>     - `node_ID_2`,

Reviewer 9: similar from another comment above but shouldn't this be `node_ID_1`? and for `node_1` above it should be `node_ID_2` since this to specify that the node knows the others public key.
